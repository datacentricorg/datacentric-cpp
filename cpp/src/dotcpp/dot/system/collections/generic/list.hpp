/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/system/ptr.hpp>
#include <dot/system/collections/list_base.hpp>

namespace dot
{
    template <class T> class ListImpl; template <class T> using List = Ptr<ListImpl<T>>;
    class TypeBuilderImpl; using TypeBuilder = Ptr<TypeBuilderImpl>;

    /// Represents a strongly typed collection of objects that can be accessed by index.
    template <class T>
    class ListImpl : public virtual ObjectImpl, public std::vector<T>, public ListBaseImpl
    {
        template <class R> friend List<R> make_list();
        template <class R> friend List<R> make_list(const std::vector<R> & obj);
        template <class R> friend List<R> make_list(std::vector<R>&& obj);
        template <class R> friend List<R> make_list(const std::initializer_list<R> & obj);
        template <class R> friend List<R> make_list(int size);

        typedef ListImpl<T> self;
        typedef std::vector<T> base;

    private: // CONSTRUCTORS

        /// Create empty list with default initial capacity.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        ListImpl() {}

        /// Construct from vector using deep copy semantics.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(const std::vector<T>& obj) : base(obj) {}

        /// Construct from vector using move semantics.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(std::vector<T>&& obj) : base(obj) {}

        /// Construct from initializer list.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(const std::initializer_list<T>& obj) : base(obj) {}

        /// Construct from int.
        ///
        /// This constructor is private. Use make_list(...) function instead.
        explicit ListImpl(int size) : base(size) {}

    public: // METHODS

        /// The number of items contained in the list.
        int count() { return this->size(); }

        /// Set total number of elements the internal data structure can hold without resizing.
        void set_capacity(int value) { this->reserve(value); }

        /// Adds an Object to the end of the list.
        void add(const T& item) { this->push_back(item); }

        /// Determines whether an element is in the list.
        virtual bool contains(const T& item)
        {
            return std::find(this->begin(), this->end(), item) != this->end();
        }

        /// Adds the elements of the specified collection to the end of the list.
        void add_range(const std::initializer_list<T>& collection)
        {
            this->insert(this->end(), collection.begin(), collection.end());
        }

        /// Adds the elements of the specified collection to the end of the list.
        void add_range(List<T> collection)
        {
            this->insert(this->end(), collection->begin(), collection->end());
        }

        /// Removes the first occurrence of a specific Object from the list.
        bool remove(const T& item)
        {
            auto iter = std::find(this->begin(), this->end(), item);
            if (iter != this->end())
            {
                this->erase(iter);
                return true;
            }
            return false;
        }

    public: // VIRTUAL METHODS

        /// Get Object from collection by index.
        virtual Object get_item(int index) override
        {
            return this->operator[](index);
        }

        /// Set Object from collection by index.
        virtual void set_item(int index, Object value) override
        {
            this->operator[](index) = (T)value;
        }

        /// Get length of collection.
        virtual int get_length() override
        {
            return this->size();
        }

        /// Add Object to end of collection.
        virtual void add_object(Object item) override
        {
            add((T)item);
        }

    public: // OPERATORS

        /// Gets or sets the element at the specified index (const version).
        virtual const T& operator[](int i) const { return base::operator[](i); }

        /// Gets or sets the element at the specified index (non-const version).
        virtual T& operator[](int i) { return base::operator[](i); }

    public: // REFLECTION

        virtual Type get_type() { return typeof(); }
        static Type typeof();
    };

    /// Create empty list with default initial capacity.
    template <class T>
    List<T> make_list() { return new ListImpl<T>(); }

    /// Construct from vector using deep copy semantics.
    template <class T>
    List<T> make_list(const std::vector<T> & obj) { return new ListImpl<T>(obj); }

    /// Construct from vector using move semantics.
    template <class T>
    List<T> make_list(std::vector<T>&& obj) { return new ListImpl<T>(std::forward<std::vector<T>>(obj)); }

    /// Construct from initializer list.
    template <class T>
    List<T> make_list(const std::initializer_list<T> & obj) { return new ListImpl<T>(obj); }

    /// Construct from int.
    template <class T>
    List<T> make_list(int size) { return new ListImpl<T>(size); }
}
