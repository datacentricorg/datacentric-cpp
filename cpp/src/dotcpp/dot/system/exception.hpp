/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/declare.hpp>

namespace dot
{
    class string;
    class string_impl;

    /// Represents errors that occur during application execution.
    class DOT_CLASS exception : public std::runtime_error
    {
        typedef std::runtime_error base;

    public: // CONSTRUCTORS

        /// Create with default message.
        exception();

        /// Create with a specified error message.
        exception(const std::string& msg);

        /// Create with a specified error message.
        exception(const char* msg);

        /// Create with a specified error message.
        exception(string msg);

    public: // METHODS

        /// Message that describes the current exception.
        virtual string message() const;
    };
}
