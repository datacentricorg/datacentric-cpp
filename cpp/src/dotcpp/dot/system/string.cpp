/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/precompiled.hpp>
#define FMT_HEADER_ONLY
#include <fmt/format.h>
#include <dot/implement.hpp>
#include <dot/system/string.hpp>
#include <dot/system/object.hpp>
#include <dot/system/nullable.hpp>
#include <dot/system/type.hpp>

namespace dot
{
    string string::empty = make_string("");

    dot::type string_impl::typeof()
    {
        static dot::type result = []()->dot::type
        {
            dot::type t = dot::make_type_builder<string_impl>("dot", "string")
                ->build();
            return t;
        }();
        return result;
    }

    dot::type string_impl::get_type()
    {
        return typeof();
    }


    bool string_impl::equals(object obj)
    {
        if (this == &(*obj)) return true;

        if (obj.is<string>())
        {
            return *this == *obj.as<string>();
        }

        return false;
    }

    size_t string_impl::hash_code()
    {
        return std::hash<std::string>()(*this);
    }

    string string_impl::to_string()
    {
        return this;
    }

    bool string_impl::ends_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(p, value->length()) == *value)
            return true;
        return false;
    }

    bool string_impl::starts_with(const string& value)
    {
        int p = length() - value->length();
        if (p >= 0 && substr(0, value->length()) == *value)
            return true;
        return false;
    }

    string string_impl::substring(int start_index)
    {
        return make_string(this->substr(start_index));
    }

    string string_impl::substring(int start_index, int length)
    {
        return make_string(this->substr(start_index, length));
    }

    int string_impl::compare_to(const string& str_b) const
    {
        return this->compare(*str_b);
    }

    int string_impl::index_of_any(list<char> any_of)
    {
        size_t pos = find_first_of(any_of->data(), 0, any_of->size());
        if (pos != std::string::npos)
            return pos;
        return -1;
    }

    string string_impl::remove(int start_index)
    {
        return make_string(*this)->erase(start_index);
    }

    string string_impl::remove(int start_index, int count)
    {
        return make_string(*this)->erase(start_index, count);
    }

    string string_impl::replace(const char old_char, const char new_char) const
    {
        string make_str = *this;
        std::replace(make_str->begin(), make_str->end(), old_char, new_char);
        return make_str;
    }

    list<string> string_impl::split(char separator) const
    {
        list<string> result = make_list<string>();

        std::size_t current, previous = 0;
        current = this->find(separator);
        while (current != std::string::npos) {
            result->add(this->substr(previous, current - previous));
            previous = current + 1;
            current = this->find(separator, previous);
        }
        result->add(this->substr(previous, current - previous));

        return result;
    }

    list<string> string_impl::split(string separator) const
    {
        list<string> result = make_list<string>();

        std::size_t current, previous = 0;
        current = this->find_first_of(*separator);
        while (current != std::string::npos) {
            result->add(this->substr(previous, current - previous));
            previous = current + 1;
            current = this->find_first_of(*separator, previous);
        }
        result->add(this->substr(previous, current - previous));

        return result;
    }

    string string_impl::trim() const
    {
        string result = make_string(*this);

        // Trim end
        result->erase(std::find_if(result->rbegin(), result->rend(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }).base(), result->end());

        // Trim start
        result->erase(result->begin(), std::find_if(result->begin(), result->end(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }));

        return result;
    }

    string string_impl::trim_start() const
    {
        string result = make_string(*this);
        result->erase(result->begin(), std::find_if(result->begin(), result->end(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }));
        return result;
    }

    string string_impl::trim_end() const
    {
        string result = make_string(*this);
        result->erase(std::find_if(result->rbegin(), result->rend(), [](char ch) {
            return !char_impl::is_white_space(ch);
        }).base(), result->end());
        return result;
    }

    bool string_impl::contains(const string& s) const
    {
        return this->find(*s) != std::string::npos;
    }

    string string_impl::to_lower() const
    {
        string result = make_string(*this);
        std::transform(result->begin(), result->end(), result->begin(),
            [](unsigned char c) { return std::tolower(c); });
        return result;
    }

    string string_impl::to_upper() const
    {
        string result = make_string(*this);
        std::transform(result->begin(), result->end(), result->begin(),
            [](unsigned char c) { return std::toupper(c); });
        return result;
    }

    bool string::is_null_or_empty(string value)
    {
        if (value == nullptr || value->empty())
            return true;
        return false;
    }

    bool string::operator==(const object& rhs) const
    {
        // If rhs is null, return false. Otherwise, check if
        // the other object is a string. If yes, compare by value.
        // If no, return false.
        if (rhs == nullptr)
        {
            return false;
        }
        else
        {
            string rhs_str = rhs.as<string>();
            if (rhs_str != nullptr) return operator==(rhs_str);
            else return false;
        }
    }

    std::string string::format_impl(fmt::string_view format_str, fmt::format_args args)
    {
        return fmt::vformat(format_str, args);
    }
}
