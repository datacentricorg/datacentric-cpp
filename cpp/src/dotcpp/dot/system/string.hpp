/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/declare.hpp>
#include <fmt/core.h>
#include <dot/detail/const_string_base.hpp>
#include <dot/system/ptr.hpp>
#include <dot/system/char.hpp>

namespace dot
{
    class string_impl; class string;
    template <class T> class list_impl; template <class T> using list = ptr<list_impl<T>>;
    template <class T> class nullable;
    enum class string_split_options;

    /// Immutable string type.
    ///
    /// The string is encoded internally as UTF-8.
    class DOT_CLASS string_impl : public virtual object_impl, public detail::ConstStringBase
    {
        typedef string_impl self;
        typedef detail::ConstStringBase base;
        friend string make_string(const std::string& rhs);
        friend string make_string(const char* rhs);

    public: // CONSTRUCTORS

        /// Creates an empty string.
        string_impl() {}

        /// Initializes from std::string or string literal.
        ///
        /// This constructor is private. Use make_string(str) function instead.
        string_impl(const std::string& value) : base(value) {}

        /// Create from const char*, null pointer is converted to to empty value.
        string_impl(const char* value) : base(value) {}

        /// Create from a single 8-bit character.
        string_impl(char value) : base(std::to_string(value)) {}

    public: // METHODS

        /// Determines whether this instance and a specified object,
        /// which must also be a string object, have the same value.
        bool equals(object obj) override;

        /// Returns the hash code for this string.
        virtual size_t hash_code() override;

        /// Returns this instance of string; no actual conversion is performed.
        virtual string to_string() override;

        /// Determines whether the end of this
        /// string matches the specified string.
        bool ends_with(const string& value);

        /// Determines whether the beginning of this
        /// string matches the specified string.
        bool starts_with(const string& value);

        /// Retrieves a substring which starts at a specified
        /// character position and continues to the end of the string.
        string substring(int start_index);

        /// Retrieves a substring which starts at the specified
        /// character position and has the specified length.
        string substring(int start_index, int length);

        /// Compares this instance with a specified string object and indicates
        /// whether this instance precedes, follows, or appears in the same position
        /// in the sort order as the specified string.
        int compare_to(const string& str_b) const;

        /// Reports the zero-based index of the first occurrence in this instance of any
        /// character in a specified array of Unicode characters.
        int index_of_any(list<char> any_of);

        /// Returns a new string in which all the characters in the current instance, beginning
        /// at a specified position and continuing through the last position, have been deleted.
        string remove(int start_index);

        /// Returns a new string in which a specified number of characters in the current
        /// instance beginning at a specified position have been deleted.
        string remove(int start_index, int count);

        /// Returns a new string in which all occurrences of a specified string
        /// in the current instance are replaced with another specified string.
        string replace(const string& old_value, const string& new_value) const; // TODO - implement

        /// Returns a new string in which all occurrences of a specified character
        /// in this instance are replaced with another specified character.
        string replace(const char old_char, const char new_char) const;

        /// Returns a string array that contains the substrings of the current string
        /// that are delimited by any of the specified 8-bit characters.
        list<string> split(char separator) const;

        /// Splits a string into substrings that are based on the characters in an array.
        list<string> split(string separator) const;

        /// Returns a string array that contains the substrings in
        /// this string that are delimited by any of the specified strings.
        /// A parameter specifies whether to return empty array elements.
        list<string> split(const list<string>& separator, const string_split_options& options) const; // TODO - implement

        /// Returns a new string in which all leading and trailing occurrences
        /// of a set of specified characters from the current string are removed.
        string trim() const;

        /// Removes all the leading occurrences of a set of characters
        /// specified in an array from the current string.
        string trim_start() const;

        /// Removes all the trailing occurrences of a set of characters
        /// specified in an array from the current string.
        string trim_end() const;

        /// Indicates whether the argument occurs within this string.
        bool contains(const string& s) const;

        /// Returns a copy of this string converted to lowercase.
        string to_lower() const;

        /// Returns a copy of this string converted to uppercase.
        string to_upper() const;

    public: // OPERATORS

        /// Returns a string containing characters from lhs followed by the characters from rhs.
        friend string operator+(const string& lhs, const string& rhs);

        /// Returns a string containing characters from lhs followed by the characters from rhs.
        friend bool operator<(const string& lhs, const string& rhs);

    public: // REFLECTION

        virtual type get_type();
        static type typeof();
    };

    /// Pointer to string_impl that has additional constructors compared to ptr(string)
    class DOT_CLASS string : public ptr<string_impl>
    {
        typedef ptr<string_impl> base;

    public: // CONSTANTS

        /// Empty string.
        static string empty;

    public: // CONSTRUCTORS

        /// Create null string.
        string() : base() {}

        /// Take ownership of raw pointer to template argument type.
        /// This also permits construction from null pointer.
        string(string_impl* p) : base(p) {}

        /// Create from std::string.
        string(const std::string& rhs) : base(new string_impl(rhs)) {}

        /// Create from string literal.
        string(const char* rhs) : base(new string_impl(rhs)) {}

        /// Copy constructor.
        string(const string& rhs) : base(rhs) {}

        /// Supports cast (explicit constructor) from object.
        ///
        /// Error if object does is not a boxed int.
        /// Null object becomes empty nullable int.
        explicit string(const ptr<object_impl>& rhs) : base(rhs) {}

    public: // STATIC

        /// Replaces format entries in the specified string with the string
        /// representation of objects in the argument array.
        template <typename ...Args>
        static string format(const string& format_string, const Args& ...args);

        /// Indicates whether the specified string is null or an string.empty string.
        static bool is_null_or_empty(string value);

    public: // OPERATORS

        /// Case sensitive comparison to std::string.
        bool operator==(const std::string& rhs) const { string_impl& impl = base::operator*(); return impl == rhs; }

        /// Case sensitive comparison to std::string.
        bool operator!=(const std::string& rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to string literal.
        bool operator==(const char* rhs) const { if (rhs == nullptr) return base::operator==(nullptr); string_impl& impl = base::operator*(); return impl == rhs; }

        /// Case sensitive comparison to string literal.
        bool operator!=(const char* rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to string literal.
        bool operator==(const ptr<string_impl>& rhs) const { return *this == *rhs; }

        /// Case sensitive comparison to string literal.
        bool operator!=(const ptr<string_impl>& rhs) const { return !operator==(rhs); }

        /// Case sensitive comparison to object.
        bool operator==(const object& rhs) const;

        /// Case sensitive comparison to object.
        bool operator!=(const object& rhs) const { return !operator==(rhs); }

    private: // STATIC

        /// Non-template implementation of string.format.
        static std::string format_impl(fmt::string_view format_str, fmt::format_args args);
    };

    /// Create from std::string or string literal using new
    inline string make_string(const std::string& rhs) { return new string_impl(rhs); }

    /// Create from std::string or string literal using new
    inline string make_string(const char* rhs) { return new string_impl(rhs); }

    /// Returns a string containing characters from lhs followed by the characters from rhs.
    inline string operator+(const string& lhs, const string& rhs) { return make_string(*lhs + *rhs); }

    /// Returns a string containing characters from lhs followed by the characters from rhs.
    inline bool operator<(const string& lhs, const string& rhs) { return *lhs < *rhs; }

    /// Helper class for fmt::format arguments conversion
    template <class T>
    struct format_forward {
        static inline auto convert(const T& t) { return t; }
    };

    /// Helper class for fmt::format arguments conversion
    template<>
    struct format_forward<string> {
        static inline std::string convert(const string& s)
        {
            if (s.is_empty())
                return std::string("");
            else
                return *s;
        }

    };

    /// Helper class for fmt::format arguments conversion
    template<class T>
    struct format_forward<ptr<T>> {
        static inline std::string convert(const ptr<T>& o) { return *o->to_string(); }
    };

    /// Helper class for fmt::format arguments conversion
    template<class T>
    struct format_forward<nullable<T>> {
        static inline T convert(const nullable<T>& o) { return o.value_or_default(); }
    };

    template<typename ...Args>
    string string::format(const string& format_string, const Args& ...args)
    {
        return format_impl(*format_string, fmt::make_format_args(format_forward<Args>::convert(args)...));
    }
}

namespace std
{
    /// Implements hash struct used by STL unordered_map for string.
    template <>
    struct hash<dot::string> : public hash<dot::ptr<dot::string_impl>>
    {};

    /// Implements equal_to struct used by STL unordered_map for string.
    template <>
    struct equal_to<dot::string> : public equal_to<dot::ptr<dot::string_impl>>
    {};
}
