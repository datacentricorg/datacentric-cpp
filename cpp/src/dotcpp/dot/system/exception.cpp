/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/precompiled.hpp>
#include <dot/implement.hpp>
#include <dot/system/exception.hpp>
#include <dot/system/string.hpp>

namespace dot
{
    exception::exception() : base("exception of type 'dot.exception' was thrown.") {}

    exception::exception(const std::string& msg) : base(msg.c_str()) {}

    exception::exception(const char* msg) : base(msg ? msg : "") {}

    exception::exception(string msg) : base(*msg) {}

    string exception::message() const { return string(what()); }
}
