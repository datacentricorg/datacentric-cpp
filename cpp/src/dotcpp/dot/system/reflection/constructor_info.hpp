/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/system/reflection/member_info.hpp>
#include <dot/system/reflection/parameter_info.hpp>
#include <dot/system/exception.hpp>

namespace dot
{
    class constructor_info_impl; using constructor_info = ptr<constructor_info_impl>;
    class type_impl; using type = ptr<type_impl>;

    /// Obtains information about the attributes of a constructor and provides access to constructor metadata.
    class constructor_info_impl : public member_info_impl
    {
        friend class type_builder_impl;

    public: // METHODS

        /// A string representing the name of the current type.
        inline virtual string to_string() override;

        /// Gets the parameters of this constructor.
        inline virtual list<parameter_info> get_parameters();

        /// Invokes specified constructor with given parameters.
        virtual object invoke(list<object>) = 0;

    protected: // CONSTRUCTORS

        list<parameter_info> parameters_;

        /// Create from declaring type
        ///
        /// This constructor is protected. It is used by derived classes only.
        inline constructor_info_impl(type declaring_type, list<attribute> custom_attributes);
    };

    /// Obtains information about the attributes of a constructor and provides access to constructor metadata.
    template <class class_, class ... args>
    class member_constructor_info_impl : public constructor_info_impl
    {

        friend class type_builder_impl;
        typedef class_(*ctor_type)(args...);

    private: // FIELDS

        ctor_type ptr_;

    public: // METHODS

        /// A string representing the name of the current type.
        inline virtual string to_string() override;

        /// Invokes the constructor reflected by this constructor_info instance.
        template <int ... I>
        inline object invoke_impl(list<object> params, detail::index_sequence<I...>);

        /// Invokes the constructor reflected by this constructor_info instance.
        inline virtual object invoke(list<object> params);

    private: // CONSTRUCTORS

        /// Create from declaring type, and pointer to constructor.
        ///
        /// This constructor is private. Use make_constructor_info(...)
        /// function with matching signature instead.
        inline member_constructor_info_impl(type declaring_type, ctor_type p, list<attribute> custom_attributes);
    };
}
