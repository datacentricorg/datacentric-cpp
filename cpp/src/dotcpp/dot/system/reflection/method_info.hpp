/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#pragma once

#include <dot/system/reflection/member_info.hpp>
#include <dot/system/reflection/parameter_info.hpp>
#include <dot/system/exception.hpp>
#include <dot/detail/traits.hpp>

namespace dot
{
    class method_info_impl; using method_info = ptr<method_info_impl>;
    class type_impl; using type = ptr<type_impl>;

    /// Obtains information about the attributes of a method and provides access to method metadata.
    class method_info_impl : public member_info_impl
    {
        friend class type_builder_impl;

        typedef method_info_impl self;

    public: // METHODS

        /// A string representing the name of the current type.
        inline virtual string to_string() override;

        /// Gets the parameters of this method.
        inline virtual list<parameter_info> get_parameters();

        /// Invokes specified method with given parameters.
        virtual object invoke(object, list<object>) = 0;

        /// Gets the return type of this method.
        type return_type; // TODO - convert to method

    protected: // FIELDS

        list<parameter_info> parameters;

    protected: // CONSTRUCTORS

        /// Create from method name, declaring type, return type.
        ///
        /// This constructor is protected. It is used by derived classes only.
        inline method_info_impl(const string& name, type declaring_type, type return_type, list<attribute> custom_attributes);
    };

    /// Obtains information about the attributes of a non-static method and provides access to method metadata.
    template <class class_, class return_t, class ... args>
    class member_method_info_impl : public method_info_impl
    {
        friend class type_builder_impl;
        typedef return_t (class_::*method_type)(args...);

    private: // FIELDS

        /// C++ function pointer type for the method.
        method_type ptr_;

    public: // METHODS

        /// A string representing the name of the current type.
        inline virtual string to_string() override;

        /// Invokes the method reflected by this method_info instance.
        template <int ... I>
        inline object invoke_impl(object obj, list<object> params, detail::index_sequence<I...>, std::false_type);

        /// Invokes the method reflected by this method_info instance.
        template <int ... I>
        inline object invoke_impl(object obj, list<object> params, detail::index_sequence<I...>, std::true_type);

        /// Invokes the method reflected by this method_info instance.
        inline virtual object invoke(object obj, list<object> params);

    private: // CONSTRUCTORS

        /// Create from method name, declaring type, return type, and pointer to method.
        ///
        /// This constructor is private. Use make_method_info(...)
        /// function with matching signature instead.
        inline member_method_info_impl(const string& name, type declaring_type, type return_type, method_type p, list<attribute> custom_attributes);
    };

    /// Obtains information about the attributes of a static method and provides access to method metadata.
    template <class return_t, class ... args>
    class static_method_info_impl : public method_info_impl
    {
        friend class type_builder_impl;
        typedef return_t (*method_type)(args...);

    private: // FIELDS

        method_type ptr_;

    public: // METHODS

        /// A string representing the name of the current type.
        virtual string to_string() override;

        /// Invokes the method reflected by this method_info instance.
        template <int ... I>
        inline object invoke_impl(object obj, list<object> params, detail::index_sequence<I...>, std::false_type);

        /// Invokes the method reflected by this method_info instance.
        template <int ... I>
        inline object invoke_impl(object obj, list<object> params, detail::index_sequence<I...>, std::true_type);

        /// Invokes the method reflected by this method_info instance.
        inline virtual object invoke(object obj, list<object> params);

    private: // CONSTRUCTORS

        /// Create from method name, declaring type, return type, and pointer to method.
        ///
        /// This constructor is private. Use make_method_info(...)
        /// function with matching signature instead.
        inline static_method_info_impl(const string& name, type declaring_type, type return_type, method_type p, list<attribute> custom_attributes);
    };
}
