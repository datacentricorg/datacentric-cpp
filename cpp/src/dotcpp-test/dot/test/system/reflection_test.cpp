/*
Copyright (C) 2015-present The DotCpp Authors.

This file is part of .C++, a native C++ implementation of
popular .NET class library APIs developed to facilitate
code reuse between C# and C++.

    http://github.com/dotcpp/dotcpp (source)
    http://dotcpp.org (documentation)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include <dot/test/implement.hpp>
#include <approvals/ApprovalTests.hpp>
#include <approvals/Catch.hpp>
#include <dot/system/string.hpp>
#include <dot/system/exception.hpp>
#include <dot/system/reflection/method_info.hpp>
#include <dot/system/reflection/constructor_info.hpp>
#include <dot/system/type.hpp>
#include <dot/system/collections/generic/list.hpp>

namespace dot
{
    static std::stringstream received;

    class reflection_base_sample_impl : public virtual object_impl
    {
        typedef reflection_base_sample_impl self;

    public: // FIELDS

        int int_field;
        int count;
        double double_field;
        list<double> double_list_field_field;

    private:

        int private_int_field = 42;

    public: // METHODS

        int sample_method(int param)
        {
            received << "invoked reflection_base_sample.sample_method";
            return 42 + param;
        }

        type get_type() override
        {
            // Create type object with thread safety guarantee as per C++ Standard
            static type result = []()->type
            {
                received << "creating type object (this should run only once)." << std::endl;

                return make_type_builder<reflection_base_sample_impl>("dot", "reflection_base_sample")
                    ->with_field("int_field", &reflection_base_sample_impl::int_field)
                    ->with_field("private_int_field", &reflection_base_sample_impl::private_int_field)
                    ->with_field("count", &reflection_base_sample_impl::count)
                    ->with_method("sample_method", &reflection_base_sample_impl::sample_method, { "param" })
                    ->build();
            }();

            return result;
        }
    };

    using reflection_base_sample = ptr<reflection_base_sample_impl>;
    reflection_base_sample make_reflection_base_sample() { return new reflection_base_sample_impl; }

    class reflection_derived_sample_impl : public reflection_base_sample_impl
    {
    public: // FIELDS

        int int_field_in_derived_class;
    };

    using reflection_derived_sample = ptr<reflection_derived_sample_impl>;
    reflection_derived_sample make_reflection_derived_sample() { return new reflection_derived_sample_impl; }

    TEST_CASE("property_info")
    {
        reflection_base_sample obj = make_reflection_base_sample();
        obj->int_field = 15;
        obj->count = 15;

        object x = obj->count;

        type result = obj->get_type();
        list<field_info> props = result->get_fields();
        field_info int_prop = props[0];
        REQUIRE(int_prop->name == "int_field");
        REQUIRE(int(int_prop->get_value(obj)) == 15);

        int_prop->set_value(obj, 19);
        REQUIRE(obj->int_field == 19);
        REQUIRE(int(int_prop->get_value(obj)) == 19);

        field_info private_int_prop = props[1];
        REQUIRE(private_int_prop->name == "private_int_field");
        REQUIRE(int(private_int_prop->get_value(obj)) == 42);

        props[2]->set_value(obj, 2384);
        REQUIRE(obj->count == 2384);
        REQUIRE(int(props[2]->get_value(obj)) == 2384);

        reflection_derived_sample obj2 = make_reflection_derived_sample();

        props[2]->set_value(obj2, -15);
        REQUIRE(obj2->count == -15);
        REQUIRE(int(props[2]->get_value(obj2)) == -15);

        list<object> params = make_list<object>(1);
        params[0] = 15;
        REQUIRE(int(result->get_methods()[0]->invoke(obj2, params)) == 42 + 15);

        Approvals::verify(received.str());
        received.clear();
    }
}
