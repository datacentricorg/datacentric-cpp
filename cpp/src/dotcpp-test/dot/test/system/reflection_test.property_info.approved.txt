creating type object (this should run only once).
invoked reflection_base_sample.sample_method
